### PR TITLE
Fixed inconsistency in gate names between text and matplotlib backends

### LINKED ISSUE #5359: Inconsistency in gate names with matplotlib and text drawer backends
### Information

- **Qiskit Terra version**: 0.23.0
- **Python version**: 3.8.6
- **Operating system**: Artix Linux

### What is the current behavior?
When drawing the complete circuit with a user-defined name, say "gate", then when drawing with the `text` backend, the circuit renders the name "gate" on the relevant portion of the circuit. However, when drawing with `matplotlib`, the particular portion of the circuit shows "Gate" i.e. it capitalizes the first character of the name.

### Steps to reproduce the problem
For example,

```python
from qiskit import *
import matplotlib.pyplot as plt

sub = QuantumCircuit(2, name='gate')
sub.h(0)
sub.cx(0,1)
sub.cz(1,0)
sub_inst = sub.to_instruction()

qc = QuantumCircuit(3)
qc.x(0)
qc.append(sub_inst, [1,2])
qc.draw('text')
```

**Output**:
```
       ┌───┐
q_0: ──┤ X ├──
     ┌─┴───┴─┐
q_1: ┤0      ├
     │  gate │
q_2: ┤1      ├
     └───────┘
```

Doing the same with `matplotlib`
```python
qc = QuantumCircuit(3)
qc.x(0)
qc.append(sub_inst, [1,2])
qc.draw('mpl')
```

**Output**:
Figure_1

### What is the expected behavior?
Both the backends must either render the name of the gate as defined by the user i.e. case-sensitive or both must capitalize the name. In my opinion, the former seems reasonable.

### PR DESCRIPTION
**Pull request by paniash**
### Summary
Fixes #5359.

### CODE REVIEW

File: qiskit/visualization/matplotlib.py [hunk 1]
```diff
@@ -367,10 +367,10 @@ def _get_gate_ctrl_text(self, op):
         if gate_text in self._style['disptex']:
             gate_text = "{}".format(self._style['disptex'][gate_text])
         else:
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:369]:
```suggestion
        elif gate_text == op.name or gate_text == base_name:
            gate_text = gate_text.capitalize()
```
For the internal op.name and base_name, we have traditionally capitalized these in the drawers, but for user-assigned labels, it makes sense to leave the case as the user chooses.

File: qiskit/visualization/matplotlib.py [hunk 2]
```diff
@@ -367,10 +367,10 @@ def _get_gate_ctrl_text(self, op):
...
-            gate_text = "{}".format(gate_text[0].upper() + gate_text[1:])
+            gate_text = "{}".format(gate_text)
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:370]:
```suggestion
```
The format statements are no longer required here.

File: qiskit/visualization/matplotlib.py [hunk 3]
```diff
@@ -367,10 +367,10 @@ def _get_gate_ctrl_text(self, op):
...
 
         if ctrl_text:
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:372]:
```suggestion
```

File: qiskit/visualization/matplotlib.py [hunk 4]
```diff
@@ -367,10 +367,10 @@ def _get_gate_ctrl_text(self, op):
...
-            ctrl_text = "{}".format(ctrl_text[0].upper() + ctrl_text[1:])
+            ctrl_text = "{}".format(ctrl_text)
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:373]:
```suggestion
```
or here.

**Comment by paniash** [comment]
@enavarro51 Hi! I've implemented the changes as per your suggestions. Let me know if it is ok. :)

File: qiskit/visualization/matplotlib.py [hunk 5]
```diff
@@ -366,11 +366,9 @@ def _get_gate_ctrl_text(self, op):
 
         if gate_text in self._style['disptex']:
             gate_text = "{}".format(self._style['disptex'][gate_text])
-        else:
-            gate_text = "{}".format(gate_text[0].upper() + gate_text[1:])
+        elif gate_text == op.name or gate_text == base_name:
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:369]:
```suggestion
        elif gate_text in (op.name, base_name):
```
One last thing to fix the lint issue. (I sometimes revert to my C language past :))

Comment (paniash) [reply qiskit/visualization/matplotlib.py]: Done!

**Comment by Cryoris** [comment]
Correct me if I'm wrong but this doesn't resolve the original issue, right?

The image below is generated using this PR's branch:

```python
>>> from qiskit import QuantumCircuit
>>> from matplotlib.pyplot import show
>>> qc = QuantumCircuit(2, name='gate')
>>> gate = qc.to_gate()
>>> qc = QuantumCircuit(3)
>>> qc.x(0)
>>> qc.initialize([0,1], 0)
>>> qc.iswap(0, 1)
>>> qc.append(gate, [1,2])
>>> qc.draw(output='mpl')
>>> show()

```
image

**Comment by enavarro51** [comment]
You're correct. Got focused on labels and missed to_gate and to_instruction names. Can be fixed by making line 369
`  elif gate_text in (op.name, base_name) and type(op.op) not in {Gate, Instruction}:`
and importing Gate and Instruction at the top.

**Comment by paniash** [comment]
@Cryoris Thank you for reporting this! I have fixed the issue as per @enavarro51's suggestion. Let me know if it's alright and feel free to report any other test cases if you fine any. ;)

**Comment by enavarro51** [comment]
I don't think there are tests for this. The easiest thing to do would be to change the `name` in `test_ghz_to_gate `to start with 'this' instead of 'This', and change a couple of the labels in `test_ctrl_labels `to start lower case. Then regenerate the reference images for these 2.

**Comment by paniash** [comment]
@enavarro51 So, I'm guessing the changes are alright for now?

**Comment by enavarro51** [comment]
No, you need to do the test part as well.
- On your local branch, from `qiskit-terra`, go to `test/mpl/ipynb/test_circuit_matplotlib_drawer.py` and make the changes I mentioned.
- Run the test file and it will create new png files in that same directory
- Copy `ctrl_labels.png` and `ghz_to_gate.png` to the references directory. Actually just checked and `big_gates.png` will also need to be copied to references.
- Push a new commit.
- You can see the results at https://mybinder.org/v2/gh/paniash/qiskit-terra/case-sensitive?urlpath=apps/test/ipynb/mpl_tester.ipynb. It will take several minutes to load. If it's all working it will be all green. You can click on any name to see the test and reference images.

Finally to get rid of the last lint, in that same line 369 from above, change `type(op.op) not in {Gate, Instruction}` to `not isinstance(op.op, (Gate, Instruction))`

File: qiskit/visualization/matplotlib.py [hunk 6]
```diff
@@ -366,11 +367,9 @@ def _get_gate_ctrl_text(self, op):
 
         if gate_text in self._style['disptex']:
             gate_text = "{}".format(self._style['disptex'][gate_text])
-        else:
-            gate_text = "{}".format(gate_text[0].upper() + gate_text[1:])
+        elif gate_text in (op.name, base_name) and isinstance(op.op, (Gate, Instruction)):
```
Comment (enavarro51) [inline qiskit/visualization/matplotlib.py:370]: should be
```suggestion
        elif gate_text in (op.name, base_name) and not isinstance(op.op, (Gate, Instruction)):
```

**Comment by paniash** [comment]
>     * You can see the results at https://mybinder.org/v2/gh/paniash/qiskit-terra/case-sensitive?urlpath=apps/test/ipynb/mpl_tester.ipynb. It will take several minutes to load. If it's all working it will be all green. You can click on any name to see the test and reference images.

@enavarro51 So, I've implemented the changes as you suggested but in the binder, I still see certain attributes to be red. Not sure what's going on.

**Review by 1ucian0 — APPROVED** [review]

**Pull request merged**